File: backend/src/main/java/com/acme/salary/salary/SalaryService.java
package com.acme.salary.salary;

import java.math.BigDecimal;
import java.math.RoundingMode;
import java.time.Clock;
import java.time.LocalDate;
import java.util.List;

import com.acme.salary.common.error.ConflictException;
import com.acme.salary.common.error.FieldValidationException;
import com.acme.salary.employee.Employee;
import com.acme.salary.employee.EmployeeNotFoundException;
import com.acme.salary.employee.EmployeeRepository;
import com.acme.salary.employee.EmploymentStatus;
import com.acme.salary.salary.dto.CorrectSalaryRequest;
import com.acme.salary.salary.dto.CreateSalaryRequest;
import com.acme.salary.salary.dto.SalaryRecordResponse;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

/**
 * Salary history for one employee at a time. "Today" comes from the injected {@link Clock}
 * (UTC), so date-dependent rules are testable.
 */
@Service
@Transactional(readOnly = true)
public class SalaryService {

	/** How far ahead a salary change may be scheduled. */
	static final int MAX_YEARS_AHEAD = 1;

	private final SalaryRecordRepository salaryRecordRepository;

	private final EmployeeRepository employeeRepository;

	private final Clock clock;

	public SalaryService(SalaryRecordRepository salaryRecordRepository, EmployeeRepository employeeRepository,
			Clock clock) {
		this.salaryRecordRepository = salaryRecordRepository;
		this.employeeRepository = employeeRepository;
		this.clock = clock;
	}

	/**
	 * The salary in force today: the record with the latest effective date on or before today.
	 * Future-dated records are ignored until their date arrives.
	 *
	 * @throws EmployeeNotFoundException if the employee does not exist
	 * @throws SalaryRecordNotFoundException if no salary is in force yet
	 */
	public SalaryRecordResponse currentSalary(long employeeId) {
		requireEmployee(employeeId);
		LocalDate today = LocalDate.now(clock);
		return salaryRecordRepository
			.findFirstByEmployeeIdAndEffectiveDateLessThanEqualOrderByEffectiveDateDesc(employeeId, today)
			.map(SalaryRecordResponse::from)
			.orElseThrow(() -> SalaryRecordNotFoundException.noCurrentSalary(employeeId, today));
	}

	/** Every salary record for the employee, newest effective date first; empty if none. */
	public List<SalaryRecordResponse> history(long employeeId) {
		requireEmployee(employeeId);
		return salaryRecordRepository.findByEmployeeIdOrderByEffectiveDateDesc(employeeId)
			.stream()
			.map(SalaryRecordResponse::from)
			.toList();
	}

	/**
	 * Adds a salary change. Earlier records are never touched, so history is preserved.
	 * <p>
	 * The duplicate-date check gives a clear 409 in the normal case. The unique
	 * {@code (employee_id, effective_date)} constraint remains the final guard: if two
	 * requests race past the check, the second insert fails and is also reported as 409.
	 */
	@Transactional
	public SalaryRecordResponse create(long employeeId, CreateSalaryRequest request) {
		Employee employee = employeeRepository.findById(employeeId)
			.orElseThrow(() -> new EmployeeNotFoundException(employeeId));
		if (employee.getEmploymentStatus() == EmploymentStatus.TERMINATED) {
			throw new ConflictException(
					"Employee " + employeeId + " is terminated; new salary records cannot be added.");
		}
		validateEffectiveDate(request.effectiveDate(), employee.getHireDate());
		if (salaryRecordRepository.existsByEmployeeIdAndEffectiveDate(employeeId, request.effectiveDate())) {
			throw new ConflictException("Employee " + employeeId + " already has a salary record effective on "
					+ request.effectiveDate() + ". Correct that record instead.");
		}

		SalaryRecord record = new SalaryRecord(employeeId, money(request.amount()), request.currency(),
				request.effectiveDate(), clock.instant());
		return SalaryRecordResponse.from(salaryRecordRepository.saveAndFlush(record));
	}

	/**
	 * Corrects the amount and currency of an existing record in place. The effective date,
	 * employee and creation time never change, and no other record is affected. Corrections
	 * are allowed for terminated employees, since they fix historical data.
	 *
	 * @throws SalaryRecordNotFoundException if the record does not exist or belongs to
	 *         another employee
	 */
	@Transactional
	public SalaryRecordResponse correct(long employeeId, long salaryId, CorrectSalaryRequest request) {
		requireEmployee(employeeId);
		SalaryRecord record = salaryRecordRepository.findByIdAndEmployeeId(salaryId, employeeId)
			.orElseThrow(() -> SalaryRecordNotFoundException.forRecord(employeeId, salaryId));
		record.correct(money(request.amount()), request.currency());
		return SalaryRecordResponse.from(salaryRecordRepository.saveAndFlush(record));
	}

	private void requireEmployee(long employeeId) {
		if (!employeeRepository.existsById(employeeId)) {
			throw new EmployeeNotFoundException(employeeId);
		}
	}

	private void validateEffectiveDate(LocalDate effectiveDate, LocalDate hireDate) {
		if (effectiveDate.isBefore(hireDate)) {
			throw new FieldValidationException("effectiveDate",
					"must not be before the employee's hire date (" + hireDate + ")");
		}
		LocalDate latestAllowed = LocalDate.now(clock).plusYears(MAX_YEARS_AHEAD);
		if (effectiveDate.isAfter(latestAllowed)) {
			throw new FieldValidationException("effectiveDate",
					"must be at most " + MAX_YEARS_AHEAD + " year in the future (" + latestAllowed + ")");
		}
	}

	// Stored as DECIMAL(15,2); validation already guarantees at most 2 decimal places.
	private static BigDecimal money(BigDecimal amount) {
		return amount.setScale(2, RoundingMode.UNNECESSARY);
	}

}

File: frontend/src/app/core/api/salary.service.ts
import { HttpClient, HttpErrorResponse } from '@angular/common/http';
import { Injectable, inject } from '@angular/core';
import { Observable, catchError, of, throwError } from 'rxjs';

import { environment } from '../../../environments/environment';
import { CorrectSalaryRequest, CreateSalaryRequest, SalaryRecord } from '../models/salary.models';

/**
 * Salary history of one employee. There is no delete: salary records are never removed.
 * The backend decides which record is current; the UI never works it out itself.
 */
@Injectable({ providedIn: 'root' })
export class SalaryService {
  private readonly http = inject(HttpClient);

  /**
   * The salary in force today, or null when none has taken effect yet. Callers load the
   * employee separately, so a 404 here means "no current salary", not "no employee".
   */
  currentSalary(employeeId: number): Observable<SalaryRecord | null> {
    return this.http.get<SalaryRecord>(this.url(employeeId)).pipe(
      catchError((error: unknown) =>
        error instanceof HttpErrorResponse && error.status === 404 ? of(null) : throwError(() => error),
      ),
    );
  }

  /** All records, newest effective date first (backend order). */
  history(employeeId: number): Observable<SalaryRecord[]> {
    return this.http.get<SalaryRecord[]>(`${this.url(employeeId)}/history`);
  }

  create(employeeId: number, request: CreateSalaryRequest): Observable<SalaryRecord> {
    return this.http.post<SalaryRecord>(this.url(employeeId), request);
  }

  /** Changes amount and currency only; the effective date is never sent. */
  correct(employeeId: number, salaryId: number, request: CorrectSalaryRequest): Observable<SalaryRecord> {
    const body: CorrectSalaryRequest = { amount: request.amount, currency: request.currency };
    return this.http.put<SalaryRecord>(`${this.url(employeeId)}/${salaryId}`, body);
  }

  private url(employeeId: number): string {
    return `${environment.apiBaseUrl}/employees/${employeeId}/salary`;
  }
}
